File: my-folder/0235-lowest-common-ancestor-of-a-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root.val == p.val or root.val == q.val:
            return root

        if root.val < p.val and root.val < q.val: 
            return self.lowestCommonAncestor(root.right, p, q)
        elif q.val < root.val and p.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return root

# [6,2,8,0,4,7,9,null,null,3,5]
# p = 2
# q = 4
# root = 6

File: my-folder/0235-lowest-common-ancestor-of-a-binary-search-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // find min of two nodes if its greater than root
        int minNode = Math.min(p.val, q.val);
        // find max btw two nodes if its less than root search lst
        int maxNode = Math.max(p.val, q.val);

        if(root.val == p.val || q.val == root.val){
            return root;
        }

        if(minNode > root.val){
            return lowestCommonAncestor(root.right, p, q);
        }
        else if(maxNode <= root.val){
            return lowestCommonAncestor(root.left, p, q);
        }
        else{
            return root;
        }
         
    }
}
